File: jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/local/LocalLambdaRunSettingsEditorPanel.java
// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

package software.aws.toolkits.jetbrains.services.lambda.local;

import com.intellij.json.JsonLanguage;
import com.intellij.openapi.components.ServiceManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.ComboBox;
import com.intellij.openapi.ui.TextFieldWithBrowseButton;
import com.intellij.ui.ComboboxWithBrowseButton;
import com.intellij.ui.EditorTextField;
import com.intellij.ui.EditorTextFieldProvider;
import com.intellij.ui.SortedComboBoxModel;
import com.intellij.util.textCompletion.TextCompletionProvider;
import com.intellij.util.textCompletion.TextFieldWithCompletion;
import com.intellij.util.ui.UIUtil;
import java.util.Collections;
import java.util.Comparator;
import javax.swing.JPanel;
import javax.swing.JRadioButton;
import software.amazon.awssdk.services.lambda.model.Runtime;
import software.aws.toolkits.core.lambda.LambdaSampleEvent;
import software.aws.toolkits.jetbrains.ui.EnvironmentVariablesTextField;

public final class LocalLambdaRunSettingsEditorPanel {
    JPanel panel;
    EditorTextField handler;
    EditorTextField inputText;
    EnvironmentVariablesTextField environmentVariables;
    ComboBox<Runtime> runtime;
    ComboboxWithBrowseButton inputTemplates;
    JRadioButton useInputFile;
    JRadioButton useInputText;
    TextFieldWithBrowseButton inputFile;
    ComboBox<LambdaSampleEvent> eventComboBox;
    SortedComboBoxModel<LambdaSampleEvent> eventComboBoxModel;
    private final Project project;
    private final TextCompletionProvider handlerCompletionProvider;

    public LocalLambdaRunSettingsEditorPanel(Project project, TextCompletionProvider handlerCompletionProvider) {
        this.project = project;
        this.handlerCompletionProvider = handlerCompletionProvider;

        useInputText.addActionListener(e -> updateComponents());
        useInputFile.addActionListener(e -> updateComponents());

        updateComponents();
    }

    private void createUIComponents() {
        EditorTextFieldProvider textFieldProvider = ServiceManager.getService(project, EditorTextFieldProvider.class);
        inputText = textFieldProvider.getEditorField(JsonLanguage.INSTANCE, project, Collections.emptyList());
        handler = new TextFieldWithCompletion(project, handlerCompletionProvider, "", true, true, true, true);

        eventComboBoxModel = new SortedComboBoxModel<>(Comparator.comparing(LambdaSampleEvent::getName));
        eventComboBox = new ComboBox<>(eventComboBoxModel);
        inputTemplates = new ComboboxWithBrowseButton(eventComboBox);
    }

    private void updateComponents() {
        inputTemplates.setEnabled(useInputText.isSelected());
        inputText.setEnabled(useInputText.isSelected());
        if(inputText.isEnabled()) {
            inputText.setBackground(UIUtil.getTextFieldBackground());
        } else {
            inputText.setBackground(panel.getBackground());
        }

        inputFile.setEnabled(useInputFile.isSelected());
    }

    public boolean isUsingInputFile() {
        return useInputFile.isSelected();
    }

    public void setUsingInputFile(boolean value) {
        useInputFile.setSelected(value);
        useInputText.setSelected(!value);

        updateComponents();
    }
}

File: plugins/toolkit/jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/execution/local/LocalLambdaRunSettingsEditorPanel.kt
// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
package software.aws.toolkits.jetbrains.services.lambda.execution.local

import com.intellij.openapi.application.runInEdt
import com.intellij.openapi.project.Project
import com.intellij.ui.IdeBorderFactory
import com.intellij.ui.components.JBRadioButton
import com.intellij.ui.components.panels.Wrapper
import com.intellij.util.ui.JBUI
import software.aws.toolkits.jetbrains.services.lambda.execution.LambdaInputPanel
import software.aws.toolkits.resources.message
import javax.swing.JCheckBox
import javax.swing.JPanel

class LocalLambdaRunSettingsEditorPanel(private val project: Project) {
    lateinit var panel: JPanel
        private set
    lateinit var invalidator: JCheckBox
        private set
    lateinit var lambdaInputPanel: JPanel
        private set
    lateinit var lambdaInput: LambdaInputPanel
        private set
    private lateinit var raw: JBRadioButton
    private lateinit var template: JBRadioButton
    private lateinit var settings: Wrapper

    val rawSettings = RawSettings(project)
    val templateSettings = TemplateSettings(project)
    var useTemplate: Boolean
        get() = template.isSelected
        set(value) {
            if (value) {
                template.isSelected = true
                settings.setContent(templateSettings.panel)
            } else {
                raw.isSelected = true
                settings.setContent(rawSettings.panel)
            }
        }

    private fun createUIComponents() {
        lambdaInput = LambdaInputPanel(project)
    }

    init {
        template.addActionListener {
            settings.setContent(templateSettings.panel)
            invalidateConfiguration()
        }
        raw.addActionListener {
            settings.setContent(rawSettings.panel)
            invalidateConfiguration()
        }
        // Select template by default
        template.isSelected = true
        settings.setContent(templateSettings.panel)

        lambdaInputPanel.border = IdeBorderFactory.createTitledBorder(message("lambda.input.label"), false, JBUI.emptyInsets())
    }

    fun invalidateConfiguration() {
        runInEdt { invalidator.isSelected = !invalidator.isSelected }
    }
}
